bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/target/obsComp/TargetSelection.java

```diff
@@ -34,110 +36,144 @@ private TargetSelection() {}
-        env.getGroups().foreach(g -> {
-            res.add(new Selection(g, null));
-            g.getTargets().foreach(t -> res.add(new Selection(g, t)));
-        });
-        env.getUserTargets().foreach(t -> res.add(new Selection(null, t)));
-        return DefaultImList.create(res);
-    }
+        boolean isGuideGroup() {
+            return guideGroup.isDefined() && target.isEmpty();
+        }
 
-    private static final MapOp<Tuple2<Selection, Integer>, Integer> INDEX_OF = Tuple2::_2;
+        boolean isGuideProbeTarget() {
+            return guideGroup.isDefined() && target.isDefined();
+        }
 
-    public static int indexOf(final TargetEnvironment env, final SPTarget target) {
-        return toSelections(env).zipWithIndex().
-                find(tup -> target.equals(tup._1().target)).
-                map(INDEX_OF).getOrElse(NO_SELECTION.value);
+        /**
+         * Create a list of Selection from the target environment, consisting of:
+         * 1. The base;
+         * 2. A list of GuideGroups and their targets; and
+         * 3. A list of the user targets.
+         */
+        static ImList<Selection> toSelections(final TargetEnvironment env) {
+            if (env == null) return ImCollections.emptyList();
+
+            int idx = 0;
+            final List<Selection> res = new ArrayList<>();
+
+            res.add(new Selection(idx++, None.instance(), new Some<>(env.getBase())));
+            for (final GuideGroup g : env.getGroups()) {
+                final Option<GuideGroup> gOpt = new Some<>(g);
+                res.add(new Selection(idx++, gOpt, None.instance()));
+                for (final Option<SPTarget> tOpt : g.getTargets().map(Some::new)) {
+                    res.add(new Selection(idx++, gOpt, tOpt));
+                }
+            }
+            for (final Option<SPTarget> tOpt : env.getUserTargets().map(Some::new)) {
+                res.add(new Selection(idx++, None.instance(), tOpt));
+            }
+            return DefaultImList.create(res);
+        }
     }
 
-    public static int indexOf(final TargetEnvironment env, final GuideGroup grp) {
-        return toSelections(env).zipWithIndex().
-                find(tup -> grp.equals(tup._1().guideGroup)).
-                map(INDEX_OF).getOrElse(NO_SELECTION.value);
+    public static Option<Integer> indexOfTarget(final TargetEnvironment env, final SPTarget target) {
+        return Selection.toSelections(env).find(sel -> sel.target.exists(target::equals)).map(sel -> sel.index);
     }
 
-    private static Selection selectionAt(final TargetEnvironment env, final int index) {
-        if (index < 0) return null;
-        final ImList<Selection> lst = toSelections(env);
-        return (index < lst.size()) ? lst.get(index) : null;
-    }
+    public static Option<Integer> indexOfIndexedGuideGroup(final TargetEnvironment env,
+                                                           final IndexedGuideGroup igg) {
```

Comment:
I may not be following this one.  You have an `IndexedGuideGroup` which pairs a `GuideGroup` with its index amongst all guide groups.  Now you want to find the index of the group in the listing including all targets and groups? Why wouldn't we just ask for the index of a `GuideGroup` regardless of where we think it is amongst its guide group peers?